pace-util/pace/util/checkpointer/thresholds.py

```diff
@@ -0,0 +1,155 @@
+import collections
+import contextlib
+import dataclasses
+from typing import Dict, List, Mapping, Union
+
+import gt4py.storage
+import numpy as np
+
+from ..quantity import Quantity
+from .base import Checkpointer
+
+
+SavepointName = str
+VariableName = str
+ArrayLike = Union[Quantity, gt4py.storage.Storage, np.ndarray]
+
+
+class InsufficientTrialsError(Exception):
+    pass
+
+
+@dataclasses.dataclass
+class Threshold:
+    relative: float
+    absolute: float
+
+    def merge(self, other: "Threshold") -> "Threshold":
+        """
+        Provide a threshold which is always satisfied
+        if both input thresholds are satisfied.
+
+        This is generally a less strict threshold than either input.
+        """
+        return Threshold(
+            relative=max(self.relative, other.relative),
+            absolute=max(self.absolute, other.absolute),
+        )
+
+
+@dataclasses.dataclass
+class SavepointThresholds:
+    savepoints: Dict[SavepointName, List[Dict[VariableName, Threshold]]]
+
+
+def cast_to_ndarray(array: ArrayLike) -> np.ndarray:
```

Comment:
If you know of multiple places where we're doing this it would be good to consolidate them to call a public helper function, but I don't want to make this function public before it's actually used that way to keep the public APIs minimal.